pkg/auth/zcapld/service.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+// Package zcapld provides zcapld service.
+package zcapld
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/hyperledger/aries-framework-go/pkg/kms"
+	"github.com/hyperledger/aries-framework-go/pkg/kms/localkms"
+	"github.com/igor-pavlenko/httpsignatures-go"
+	"github.com/trustbloc/edge-core/pkg/log"
+	"github.com/trustbloc/edge-core/pkg/zcapld"
+)
+
+const (
+	signEndpoint = "/keys/%s/sign"
+)
+
+var logger = log.New("auth-service")
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// Service to provide zcapld functionality.
+type Service struct {
+	authzKeyStoreURL string
+	httpClient       httpClient
+}
+
+type signReq struct {
+	Message string `json:"message,omitempty"`
+}
+
+type signResp struct {
+	Signature string `json:"signature,omitempty"`
+}
+
+// New return zcap service.
+func New(authzKeyStoreURL string) *Service {
+	return &Service{authzKeyStoreURL: authzKeyStoreURL, httpClient: &http.Client{}}
+}
+
+// SignHeader sign header.
+func (s *Service) SignHeader(req *http.Request, capabilityBytes []byte) (*http.Header, error) {
+	capability, err := zcapld.ParseCapability(capabilityBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	compressedZcap, err := compressZCAP(capability)
+	if err != nil {
+		return nil, err
+	}
+
+	action := "write"
+	if req.Method == http.MethodGet {
+		action = "read"
+	}
+
+	req.Header.Set(zcapld.CapabilityInvocationHTTPHeader,
+		fmt.Sprintf(`zcap capability="%s",action="%s"`, compressedZcap, action))
+
+	hs := httpsignatures.NewHTTPSignatures(&zcapld.AriesDIDKeySecrets{})
+	hs.SetSignatureHashAlgorithm(&didKeySignatureHashAlgorithm{
+		s: s,
+	})
+
+	err = hs.Sign(capability.Invoker, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return &req.Header, nil
+}
+
+func compressZCAP(zcap *zcapld.Capability) (string, error) {
```

Comment:
I have seen this function multiple times - should this be part of common code in edge-core? @fqutishat @llorllale

Reply:
yes

Reply:
yes